kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,308 @@
+    these streams correspond to stdin and stdout.
+
+-   The data on each channel are partitioned into consecutive **frames**. Each
+    frame consists of a length tag followed by a message body of that length.
+
+    A **length tag** is a byte count _n_ encoded as a minimum-width decimal
+    integer using the Unicode digits `0` (code 48) through `9` (code 57), and
+    terminated by a newline (code 10). A **message body** is the sequence of _n_
+    bytes immediately following the length tag.
+
+    For example, a frame containing the message `foobar` is encoded as
+
+        6    \n  f    o    o    b    a    r      -- text
+        0x36 0xa 0x66 0x6f 0x6f 0x62 0x61 0x72   -- Unicode (hex)
+
+-   Within this framing, the analyzer communicates with the driver using
+    [JSON-RPC 2.0](http://www.jsonrpc.org/specification) requests and responses,
+    as defined below. The analyzer is the JSON-RPC client, and the driver is the
+    server.
+
+    The server MUST support concurrent JSON-RPC requests from the analyzer. The
+    server MUST support batch JSON-RPC requests from the analyzer.
+
+-   Under certain circumstances, the analyzer process **must exit**. In a case
+    where the analyzer process must exit, it MUST terminate within a grace
+    period. If it does not, the operating environment SHOULD forcibly terminate
+    the process (and may do so at any time after the grace period ends). The
+    exit code of an analyzer process is ignored. The grace period is defined by
+    the implementation, but must be at least 10 seconds.
+
+-   If and after the analyzer closes the output stream, the analyzer process
+    must exit.
+
+-   If and after the driver closes the input stream, the analyzer process must
+    exit.
+
+-   A **corrupt frame** is one in which the length tag does not have the correct
+    form. A frame is corrupt if the newline terminator is missing, the string
+    preceding the newline terminator contains non-digits or is empty, or the
+    stream ends before the specified number of bytes is delivered.
+
+    If the analyzer receives a corrupt frame from the driver, it must exit. If
+    the driver receives a corrupt frame from the analyzer, it must close the
+    input stream (in which case the analyzer therefore must exit).
+
+-   When an analyzer process terminates, the driver MUST consider any pending
+    analysis requests to have failed, and thereafter the environment MAY retry
+    analysis of the corresponding compilation records.
+
+### Messages
+
+The following message structures are used by the protocol:
+
+-   `InitRequest`
+
+    ```
+    {
+      "protocol": <protocol>
+      "output-encoding": <output-encoding>
+    }
+    ```
+
+-   `InitReply`
+
+    ```
+    {
+      "protocol": <protocol>
+    }
+    ```
+
+-   `AnalyzerRequest`
+
+    ```
+    {
+      "type": <analysis-type>
+    }
+    ```
+
+-   `AnalyzeReply`
+
+    ```
+    {
+      "id": <analysis-id>,
```

Comment:
Can the compilation unit remain as part of the `AnalyzeReply`?  Or must it be part of the FS?

Reply:
Changed so that the non-file part of the CU is in the AnalyzeReply.